lib/SPIRV/libSPIRV/SPIRVInstruction.h

```diff
@@ -2871,6 +2871,74 @@ class SPIRVExpectKHRInstBase : public SPIRVInstTemplateBase {
 _SPIRV_OP(ExpectKHR, true, 5)
 #undef _SPIRV_OP
 
+class SPIRVDotKHRBase : public SPIRVInstTemplateBase {
+protected:
+  SPIRVCapVec getRequiredCapability() const override {
+    // Both vector operands must have the same type, so analyzing the
+    // first operand will suffice.
+    SPIRVCapabilityKind ArgCap = getRequiredCapabilityForOperand(Ops[0]);
+    return getVec(ArgCap, CapabilityDotProductKHR);
+  }
+
+  llvm::Optional<ExtensionID> getRequiredExtension() const override {
+    return ExtensionID::SPV_KHR_integer_dot_product;
+  }
+
+  void validate() const override {
+    SPIRVInstruction::validate();
+    SPIRVId Vec1 = Ops[0];
+    SPIRVId Vec2 = Ops[1];
+    (void)Vec1;
+    (void)Vec2;
+
+    assert(getValueType(Vec1) == getValueType(Vec2) &&
```

Comment:
Module's error log may be used instead of assert:
getModule()->getErrorLog().checkError(getValueType(Vec1) == getValueType(Vec2), SPIRVEC_InvalidInstruction, "Input vectors must have the same type");
```suggestion
    getModule()->getErrorLog().checkError(getValueType(Vec1) == getValueType(Vec2), SPIRVEC_InvalidInstruction, OpCodeNameMap::map(OpCode) + "\nInput vectors must have the same type");
```
In this case llvm-spirv will be finished without crash on error and the check will work even in build w/o asserts.

Reply:
All SPIRVInstruction validate functions are relying on asserts, none of them is using the Module error log currently.  So I prefer not to break this pattern as part of this PR.  I agree there are many ways to improve those validate functions though.

> and the check will work even in build w/o asserts.

As I understand, these validate functions are mostly there to aid translator developers.  They do not perform full validation, as there's the separate spirv-val tool for that.  So having those checks in builds without asserts isn't that important I think.

Reply:
Ok, I see, thank you for explanation.